### ISSUE TITLE
TensorFlow Mechanics 101

### ISSUE THREAD

**Issue by killelder** [killelder/tensorflownote#5]
https://www.tensorflow.org/get_started/mnist/mechanics

---------------------------------------------------------------
這篇主要參考下面兩段code
這篇感覺是直接把MNIST的東西灌入tensorflow訓練

mnist.py
https://github.com/tensorflow/tensorflow/blob/r1.4/tensorflow/examples/tutorials/mnist/mnist.py
fully_connected_feed.py
https://github.com/tensorflow/tensorflow/blob/r1.4/tensorflow/examples/tutorials/mnist/fully_connected_feed.py

mnist.py主要是建構圖,
主要根據3個stage
1. inference() 建構圖到可以做predict的部分
2. loss() 加一個operation 計算loss 在inference上
3. training() 加在loss 上, 要可以計算而且更改weight bias

看到目前為止, 建構graph的方式大致上都跟前兩章差不多
比較特別的是使用了更明顯分層的概念,
由mnist.py定義好3個stage要做的事情
fully_connected_feed.py一層一層帶入
前大半段我就不特別解說
--------------------------------------------------------
ex. 1-1
with tf.Graph().as_default():
tf.Graph = 是整張圖包含著那些operations, 一次會執行所有相同group內的所有operations,
大部分只需要根據single default graph
要用很多張圖也可以, 不過這裡不介紹

ex. 1-2
init = tf.global_variables_initializer()
sess.run(init)

當執行init, init operation只有包含initialize 所有的variables, 其他group不會執行

ex. 1-3
for step in xrange(FLAGS.max_steps):
    sess.run(train_op)

這裡還必須要把data切片, 符合每次step的batch_size
loss function值如果出現NaN就是代表發散了

------------------------------------------------------------------
Visualize the Status
為了想在tensorboard看到所有的summaries會被收集到一個tensor
ex. 1-4
summary = tf.summary.merge_all()

在session建立以後,  利用tf.summary.FileWriter 可以輸出event file,
包含graph和summaries的值,
然後每次更新summary, 利用add_summary把它寫入檔案

ex. 1-5
summary_writer = tf.summary.FileWriter(FLAGS.train_dir, sess.graph)
summary_str = sess.run(summary, feed_dict=feed_dict)
summary_writer.add_summary(summary_str, step)

Save a checkpoint
利用tf.train.Saver去存model, 並利用save function去存檔
利用restore去讀檔

ex. 1-6
saver = tf.train.Saver()
saver.save(sess, FLAGS.train_dir, global_step=step)

ex. 1-7
saver.restore(sess, FLAGS.train_dir)

下面用到了評估model的一個方式, 看第k名預測中,
如果第3名才預測中, 那就是3

ex. 1-8
eval_correct = tf.nn.in_top_k(logits, labels, 1)

**然而最後的precision看不是很懂那個評估方式
